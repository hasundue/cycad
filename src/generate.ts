import { distinct } from "../lib/std/collections.ts";
import { $ } from "../lib/x/dax.ts";
import { ensure, is } from "../lib/x/unknownutil.ts";
import { throw_ } from "./utils.ts";

export async function fetchGrammarList() {
  const response = await fetch(
    `https://api.github.com/repos/NixOS/nixpkgs/contents/pkgs/development/tools/parsing/tree-sitter/grammars`,
  );
  const json = ensure(
    await response.json(),
    is.ArrayOf(is.ObjectOf({ name: is.String })),
  );
  return json.filter((item) => item.name.endsWith(".json"))
    .map((item) => item.name);
}

function parseGrammerName(filename: string) {
  return filename.split("-")[2].split(".")[0];
}

function createLanguageList(grammars: string[]) {
  return distinct(
    grammars.map((g) => parseGrammerName(g))
      .concat(Object.keys(LanguageSpecOverlays)),
  ).sort();
}

interface LanguageSpecOverlay {
  grammar?: string;
  language?: string;
  location?: string;
  generate?: boolean;
}

// Hard-ported from https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/parsing/tree-sitter/default.nix
const LanguageSpecOverlays: Record<string, LanguageSpecOverlay | undefined> = {
  ocaml: {
    grammar: "ocaml",
    location: "ocaml",
  },
  "ocaml-interface": {
    grammar: "ocaml",
    location: "interface",
  },
  "org-nvim": {
    language: "org",
  },
  typescript: {
    grammar: "typescript",
    location: "typescript",
  },
  tsx: {
    grammar: "typescript",
    location: "tsx",
  },
  typst: {
    generate: true,
  },
  markdown: {
    grammar: "markdown",
    location: "tree-sitter-markdown",
  },
  "markdown-inline": {
    grammar: "markdown",
    language: "markdown_inline",
    location: "tree-sitter-markdown-inline",
  },
  wing: {
    location: "libs/tree-sitter-wing",
    generate: true,
  },
};

const isGrammarJson = is.ObjectOf({
  url: is.String,
  fetchLFS: is.OptionalOf(is.LiteralOf(false)),
  fetchSubmodules: is.OptionalOf(is.LiteralOf(false)),
  deepClone: is.OptionalOf(is.LiteralOf(false)),
});

type GrammarJson = {
  grammar: string;
  url: string;
};

async function fetchGrammarJson(
  grammar: string,
): Promise<GrammarJson> {
  const response = await fetch(
    `https://raw.githubusercontent.com/NixOS/nixpkgs/master/pkgs/development/tools/parsing/tree-sitter/grammars/${grammar}`,
  );
  if (!response.ok) {
    throw new Error(
      `Could not find tree-sitter grammer for ${grammar} in nixpkgs`,
    );
  }
  const json = ensure(
    await response.json(),
    isGrammarJson,
  );
  return {
    grammar: parseGrammerName(grammar),
    url: json.url,
  };
}

async function fetchGrammarJsons(
  grammars: string[],
): Promise<GrammarJson[]> {
  const jsons: GrammarJson[] = [];
  const progbar = $.progress(
    `Fetching metadata for grammars from nixpkgs repository`,
    { length: grammars.length },
  );
  await progbar.with(async () => {
    // We can't use Promise.all because GitHub will rate limit us
    for (const grammar of grammars) {
      jsons.push(await fetchGrammarJson(grammar));
      progbar.increment();
    }
  });
  return jsons;
}

type LanguageSpec = GrammarJson & Required<Omit<LanguageSpecOverlay, "name">>;

type LanguageSpecMap = Record<string, LanguageSpec>;

export async function createLanguageSpecMap(): Promise<LanguageSpecMap> {
  const grammars = await fetchGrammarList();
  const jsons = await fetchGrammarJsons(grammars);
  const languages = createLanguageList(grammars);

  return Object.fromEntries(languages.map((language) => {
    const overlay = LanguageSpecOverlays[language];
    const grammar = overlay?.grammar ?? overlay?.language ?? language;
    const json = jsons.find((j) => j.grammar === grammar) ??
      throw_(new Error(`Could not find grammar for ${language}`));
    return [
      language,
      {
        ...json,
        language: overlay?.language ?? language,
        location: overlay?.location ?? ".",
        generate: overlay?.generate ?? false,
      },
    ];
  }));
}

async function generateLanguageSpecModule() {
  const content = `
// Generated by src/langs.ts (do not edit manually)

export const LanguageSpecMap = ${
    JSON.stringify(await createLanguageSpecMap(), null, 2)
  } as const;

export type Language = keyof typeof LanguageSpecMap;


export const isLanguage = (lang: string): lang is Language => {
  return lang in LanguageSpecMap;
};
`;
  const url = new URL("./langs.generated.ts", import.meta.url);
  await Deno.writeTextFile(
    url,
    content,
  );
  const path = $.path(url);
  await $`deno fmt ${path}`.quiet();
}

if (import.meta.main) {
  try {
    await generateLanguageSpecModule();
  } catch (error) {
    console.error(error);
    Deno.exit(1);
  }
}
