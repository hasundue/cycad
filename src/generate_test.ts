import {
  assertSnapshot,
} from "https://deno.land/std@0.204.0/testing/snapshot.ts";
import { createLanguageSpecMap } from "./generate.ts";

Deno.test("createLanguageSpecs", { ignore: true }, async (t) => {
  const specs = await createLanguageSpecMap();
  await assertSnapshot(t, specs);
});
