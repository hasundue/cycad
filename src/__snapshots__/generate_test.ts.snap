export const snapshot = {};

snapshot[`createLanguageSpecs 1`] = `
{
  "markdown-inline": {
    generate: false,
    grammar: "markdown",
    language: "markdown_inline",
    location: "tree-sitter-markdown-inline",
    url: "https://github.com/MDeiml/tree-sitter-markdown",
  },
  "ocaml-interface": {
    generate: false,
    grammar: "ocaml",
    language: "ocaml-interface",
    location: "interface",
    url: "https://github.com/tree-sitter/tree-sitter-ocaml",
  },
  "org-nvim": {
    generate: false,
    grammar: "org",
    language: "org",
    location: ".",
    url: "https://github.com/milisims/tree-sitter-org",
  },
  bash: {
    generate: false,
    grammar: "bash",
    language: "bash",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-bash",
  },
  beancount: {
    generate: false,
    grammar: "beancount",
    language: "beancount",
    location: ".",
    url: "https://github.com/polarmutex/tree-sitter-beancount",
  },
  bibtex: {
    generate: false,
    grammar: "bibtex",
    language: "bibtex",
    location: ".",
    url: "https://github.com/latex-lsp/tree-sitter-bibtex",
  },
  c: {
    generate: false,
    grammar: "c",
    language: "c",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-c-sharp",
  },
  clojure: {
    generate: false,
    grammar: "clojure",
    language: "clojure",
    location: ".",
    url: "https://github.com/sogaiu/tree-sitter-clojure",
  },
  cmake: {
    generate: false,
    grammar: "cmake",
    language: "cmake",
    location: ".",
    url: "https://github.com/uyha/tree-sitter-cmake",
  },
  comment: {
    generate: false,
    grammar: "comment",
    language: "comment",
    location: ".",
    url: "https://github.com/stsewd/tree-sitter-comment",
  },
  commonlisp: {
    generate: false,
    grammar: "commonlisp",
    language: "commonlisp",
    location: ".",
    url: "https://github.com/thehamsta/tree-sitter-commonlisp",
  },
  cpp: {
    generate: false,
    grammar: "cpp",
    language: "cpp",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-cpp",
  },
  css: {
    generate: false,
    grammar: "css",
    language: "css",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-css",
  },
  cuda: {
    generate: false,
    grammar: "cuda",
    language: "cuda",
    location: ".",
    url: "https://github.com/thehamsta/tree-sitter-cuda",
  },
  cue: {
    generate: false,
    grammar: "cue",
    language: "cue",
    location: ".",
    url: "https://github.com/eonpatapon/tree-sitter-cue",
  },
  dart: {
    generate: false,
    grammar: "dart",
    language: "dart",
    location: ".",
    url: "https://github.com/usernobody14/tree-sitter-dart",
  },
  devicetree: {
    generate: false,
    grammar: "devicetree",
    language: "devicetree",
    location: ".",
    url: "https://github.com/joelspadin/tree-sitter-devicetree",
  },
  dockerfile: {
    generate: false,
    grammar: "dockerfile",
    language: "dockerfile",
    location: ".",
    url: "https://github.com/camdencheek/tree-sitter-dockerfile",
  },
  dot: {
    generate: false,
    grammar: "dot",
    language: "dot",
    location: ".",
    url: "https://github.com/rydesun/tree-sitter-dot",
  },
  eex: {
    generate: false,
    grammar: "eex",
    language: "eex",
    location: ".",
    url: "https://github.com/connorlay/tree-sitter-eex",
  },
  elisp: {
    generate: false,
    grammar: "elisp",
    language: "elisp",
    location: ".",
    url: "https://github.com/wilfred/tree-sitter-elisp",
  },
  elixir: {
    generate: false,
    grammar: "elixir",
    language: "elixir",
    location: ".",
    url: "https://github.com/elixir-lang/tree-sitter-elixir",
  },
  elm: {
    generate: false,
    grammar: "elm",
    language: "elm",
    location: ".",
    url: "https://github.com/elm-tooling/tree-sitter-elm",
  },
  embedded: {
    generate: false,
    grammar: "embedded",
    language: "embedded",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-embedded-template",
  },
  erlang: {
    generate: false,
    grammar: "erlang",
    language: "erlang",
    location: ".",
    url: "https://github.com/abstractmachineslab/tree-sitter-erlang",
  },
  fennel: {
    generate: false,
    grammar: "fennel",
    language: "fennel",
    location: ".",
    url: "https://github.com/travonted/tree-sitter-fennel",
  },
  fish: {
    generate: false,
    grammar: "fish",
    language: "fish",
    location: ".",
    url: "https://github.com/ram02z/tree-sitter-fish",
  },
  fortran: {
    generate: false,
    grammar: "fortran",
    language: "fortran",
    location: ".",
    url: "https://github.com/stadelmanma/tree-sitter-fortran",
  },
  gdscript: {
    generate: false,
    grammar: "gdscript",
    language: "gdscript",
    location: ".",
    url: "https://github.com/prestonknopp/tree-sitter-gdscript",
  },
  glimmer: {
    generate: false,
    grammar: "glimmer",
    language: "glimmer",
    location: ".",
    url: "https://github.com/alexlafroscia/tree-sitter-glimmer",
  },
  glsl: {
    generate: false,
    grammar: "glsl",
    language: "glsl",
    location: ".",
    url: "https://github.com/thehamsta/tree-sitter-glsl",
  },
  go: {
    generate: false,
    grammar: "go",
    language: "go",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-go",
  },
  godot: {
    generate: false,
    grammar: "godot",
    language: "godot",
    location: ".",
    url: "https://github.com/prestonknopp/tree-sitter-godot-resource",
  },
  gomod: {
    generate: false,
    grammar: "gomod",
    language: "gomod",
    location: ".",
    url: "https://github.com/camdencheek/tree-sitter-go-mod",
  },
  gowork: {
    generate: false,
    grammar: "gowork",
    language: "gowork",
    location: ".",
    url: "https://github.com/omertuc/tree-sitter-go-work",
  },
  graphql: {
    generate: false,
    grammar: "graphql",
    language: "graphql",
    location: ".",
    url: "https://github.com/bkegley/tree-sitter-graphql",
  },
  haskell: {
    generate: false,
    grammar: "haskell",
    language: "haskell",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-haskell",
  },
  hcl: {
    generate: false,
    grammar: "hcl",
    language: "hcl",
    location: ".",
    url: "https://github.com/MichaHoffmann/tree-sitter-hcl",
  },
  heex: {
    generate: false,
    grammar: "heex",
    language: "heex",
    location: ".",
    url: "https://github.com/connorlay/tree-sitter-heex",
  },
  hjson: {
    generate: false,
    grammar: "hjson",
    language: "hjson",
    location: ".",
    url: "https://github.com/winston0410/tree-sitter-hjson",
  },
  html: {
    generate: false,
    grammar: "html",
    language: "html",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-html",
  },
  http: {
    generate: false,
    grammar: "http",
    language: "http",
    location: ".",
    url: "https://github.com/ntbbloodbath/tree-sitter-http",
  },
  janet: {
    generate: false,
    grammar: "janet",
    language: "janet",
    location: ".",
    url: "https://github.com/sogaiu/tree-sitter-janet-simple",
  },
  java: {
    generate: false,
    grammar: "java",
    language: "java",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-java",
  },
  javascript: {
    generate: false,
    grammar: "javascript",
    language: "javascript",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-javascript",
  },
  jsdoc: {
    generate: false,
    grammar: "jsdoc",
    language: "jsdoc",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-jsdoc",
  },
  json5: {
    generate: false,
    grammar: "json5",
    language: "json5",
    location: ".",
    url: "https://github.com/joakker/tree-sitter-json5",
  },
  json: {
    generate: false,
    grammar: "json",
    language: "json",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-json",
  },
  jsonnet: {
    generate: false,
    grammar: "jsonnet",
    language: "jsonnet",
    location: ".",
    url: "https://github.com/sourcegraph/tree-sitter-jsonnet",
  },
  julia: {
    generate: false,
    grammar: "julia",
    language: "julia",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-julia",
  },
  kotlin: {
    generate: false,
    grammar: "kotlin",
    language: "kotlin",
    location: ".",
    url: "https://github.com/fwcd/tree-sitter-kotlin",
  },
  latex: {
    generate: false,
    grammar: "latex",
    language: "latex",
    location: ".",
    url: "https://github.com/latex-lsp/tree-sitter-latex",
  },
  ledger: {
    generate: false,
    grammar: "ledger",
    language: "ledger",
    location: ".",
    url: "https://github.com/cbarrete/tree-sitter-ledger",
  },
  llvm: {
    generate: false,
    grammar: "llvm",
    language: "llvm",
    location: ".",
    url: "https://github.com/benwilliamgraham/tree-sitter-llvm",
  },
  lua: {
    generate: false,
    grammar: "lua",
    language: "lua",
    location: ".",
    url: "https://github.com/MunifTanjim/tree-sitter-lua",
  },
  make: {
    generate: false,
    grammar: "make",
    language: "make",
    location: ".",
    url: "https://github.com/alemuller/tree-sitter-make",
  },
  markdown: {
    generate: false,
    grammar: "markdown",
    language: "markdown",
    location: "tree-sitter-markdown",
    url: "https://github.com/MDeiml/tree-sitter-markdown",
  },
  nickel: {
    generate: false,
    grammar: "nickel",
    language: "nickel",
    location: ".",
    url: "https://github.com/nickel-lang/tree-sitter-nickel",
  },
  nix: {
    generate: false,
    grammar: "nix",
    language: "nix",
    location: ".",
    url: "https://github.com/cstrahan/tree-sitter-nix",
  },
  norg: {
    generate: false,
    grammar: "norg",
    language: "norg",
    location: ".",
    url: "https://github.com/nvim-neorg/tree-sitter-norg",
  },
  nu: {
    generate: false,
    grammar: "nu",
    language: "nu",
    location: ".",
    url: "https://github.com/LhKipp/tree-sitter-nu",
  },
  ocaml: {
    generate: false,
    grammar: "ocaml",
    language: "ocaml",
    location: "ocaml",
    url: "https://github.com/tree-sitter/tree-sitter-ocaml",
  },
  org: {
    generate: false,
    grammar: "org",
    language: "org",
    location: ".",
    url: "https://github.com/milisims/tree-sitter-org",
  },
  perl: {
    generate: false,
    grammar: "perl",
    language: "perl",
    location: ".",
    url: "https://github.com/ganezdragon/tree-sitter-perl",
  },
  pgn: {
    generate: false,
    grammar: "pgn",
    language: "pgn",
    location: ".",
    url: "https://github.com/rolandwalker/tree-sitter-pgn",
  },
  php: {
    generate: false,
    grammar: "php",
    language: "php",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-php",
  },
  pioasm: {
    generate: false,
    grammar: "pioasm",
    language: "pioasm",
    location: ".",
    url: "https://github.com/leo60228/tree-sitter-pioasm",
  },
  prisma: {
    generate: false,
    grammar: "prisma",
    language: "prisma",
    location: ".",
    url: "https://github.com/victorhqc/tree-sitter-prisma",
  },
  proto: {
    generate: false,
    grammar: "proto",
    language: "proto",
    location: ".",
    url: "https://github.com/mitchellh/tree-sitter-proto",
  },
  pug: {
    generate: false,
    grammar: "pug",
    language: "pug",
    location: ".",
    url: "https://github.com/zealot128/tree-sitter-pug",
  },
  python: {
    generate: false,
    grammar: "python",
    language: "python",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-python",
  },
  ql: {
    generate: false,
    grammar: "ql",
    language: "ql",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-ql-dbscheme",
  },
  query: {
    generate: false,
    grammar: "query",
    language: "query",
    location: ".",
    url: "https://github.com/nvim-treesitter/tree-sitter-query",
  },
  r: {
    generate: false,
    grammar: "r",
    language: "r",
    location: ".",
    url: "https://github.com/r-lib/tree-sitter-r",
  },
  razor: {
    generate: false,
    grammar: "razor",
    language: "razor",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-razor",
  },
  regex: {
    generate: false,
    grammar: "regex",
    language: "regex",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-regex",
  },
  rego: {
    generate: false,
    grammar: "rego",
    language: "rego",
    location: ".",
    url: "https://github.com/FallenAngel97/tree-sitter-rego",
  },
  rst: {
    generate: false,
    grammar: "rst",
    language: "rst",
    location: ".",
    url: "https://github.com/stsewd/tree-sitter-rst",
  },
  ruby: {
    generate: false,
    grammar: "ruby",
    language: "ruby",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-ruby",
  },
  rust: {
    generate: false,
    grammar: "rust",
    language: "rust",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-rust",
  },
  scala: {
    generate: false,
    grammar: "scala",
    language: "scala",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-scala",
  },
  scheme: {
    generate: false,
    grammar: "scheme",
    language: "scheme",
    location: ".",
    url: "https://github.com/6cdh/tree-sitter-scheme",
  },
  scss: {
    generate: false,
    grammar: "scss",
    language: "scss",
    location: ".",
    url: "https://github.com/serenadeai/tree-sitter-scss",
  },
  smithy: {
    generate: false,
    grammar: "smithy",
    language: "smithy",
    location: ".",
    url: "https://github.com/indoorvivants/tree-sitter-smithy",
  },
  solidity: {
    generate: false,
    grammar: "solidity",
    language: "solidity",
    location: ".",
    url: "https://github.com/JoranHonig/tree-sitter-solidity",
  },
  sparql: {
    generate: false,
    grammar: "sparql",
    language: "sparql",
    location: ".",
    url: "https://github.com/bonabeavis/tree-sitter-sparql",
  },
  sql: {
    generate: false,
    grammar: "sql",
    language: "sql",
    location: ".",
    url: "https://github.com/derekstride/tree-sitter-sql",
  },
  supercollider: {
    generate: false,
    grammar: "supercollider",
    language: "supercollider",
    location: ".",
    url: "https://github.com/madskjeldgaard/tree-sitter-supercollider",
  },
  surface: {
    generate: false,
    grammar: "surface",
    language: "surface",
    location: ".",
    url: "https://github.com/connorlay/tree-sitter-surface",
  },
  svelte: {
    generate: false,
    grammar: "svelte",
    language: "svelte",
    location: ".",
    url: "https://github.com/Himujjal/tree-sitter-svelte",
  },
  tiger: {
    generate: false,
    grammar: "tiger",
    language: "tiger",
    location: ".",
    url: "https://github.com/ambroisie/tree-sitter-tiger",
  },
  tlaplus: {
    generate: false,
    grammar: "tlaplus",
    language: "tlaplus",
    location: ".",
    url: "https://github.com/tlaplus-community/tree-sitter-tlaplus",
  },
  toml: {
    generate: false,
    grammar: "toml",
    language: "toml",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-toml",
  },
  tsq: {
    generate: false,
    grammar: "tsq",
    language: "tsq",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-tsq",
  },
  tsx: {
    generate: false,
    grammar: "typescript",
    language: "tsx",
    location: "tsx",
    url: "https://github.com/tree-sitter/tree-sitter-typescript",
  },
  turtle: {
    generate: false,
    grammar: "turtle",
    language: "turtle",
    location: ".",
    url: "https://github.com/bonabeavis/tree-sitter-turtle",
  },
  typescript: {
    generate: false,
    grammar: "typescript",
    language: "typescript",
    location: "typescript",
    url: "https://github.com/tree-sitter/tree-sitter-typescript",
  },
  typst: {
    generate: true,
    grammar: "typst",
    language: "typst",
    location: ".",
    url: "https://github.com/uben0/tree-sitter-typst",
  },
  verilog: {
    generate: false,
    grammar: "verilog",
    language: "verilog",
    location: ".",
    url: "https://github.com/tree-sitter/tree-sitter-verilog",
  },
  vim: {
    generate: false,
    grammar: "vim",
    language: "vim",
    location: ".",
    url: "https://github.com/vigoux/tree-sitter-viml",
  },
  vue: {
    generate: false,
    grammar: "vue",
    language: "vue",
    location: ".",
    url: "https://github.com/ikatyang/tree-sitter-vue",
  },
  wing: {
    generate: true,
    grammar: "wing",
    language: "wing",
    location: "libs/tree-sitter-wing",
    url: "https://github.com/winglang/wing",
  },
  yaml: {
    generate: false,
    grammar: "yaml",
    language: "yaml",
    location: ".",
    url: "https://github.com/ikatyang/tree-sitter-yaml",
  },
  yang: {
    generate: false,
    grammar: "yang",
    language: "yang",
    location: ".",
    url: "https://github.com/hubro/tree-sitter-yang",
  },
  zig: {
    generate: false,
    grammar: "zig",
    language: "zig",
    location: ".",
    url: "https://github.com/maxxnino/tree-sitter-zig",
  },
}
`;
